h5py/_conv.pyx

```diff
@@ -714,17 +716,30 @@ cdef int conv_vlen2ndarray(void* ipt,
     itemsize = H5Tget_size(outtype.id)
     if itemsize > H5Tget_size(intype.id):
         data = realloc(data, itemsize * size)
-    H5Tconvert(intype.id, outtype.id, size, data, NULL, H5P_DEFAULT)
+
+    if needs_bkg_buffer(intype.id, outtype.id):
+        back_buf = malloc(H5Tget_size(outtype.id)*size)
+
+    H5Tconvert(intype.id, outtype.id, size, data, back_buf, H5P_DEFAULT)
+
+    # We need to use different approaches to creating the ndarray with the converted
+    # data depending on the destination dtype.
+    # For simple dtypes, we can use SimpleNewFromData, but types like
+    # string & void need a size specified, so this function can't be used.
+    # Additionally, Cython doesn't expose NumPy C-API functions like NewFromDescr,
+    # so we fall back on the Python frombuffer() function. Finally, that function does
+    # not support object arrays, so in that case we copy directly to the underlying buffer
+    # of a new ndarray.
 
     if elem_dtype.kind in b"biufcmMO":
         # type_num is enough to create an array for these dtypes
         ndarray = cnp.PyArray_SimpleNewFromData(1, dims, elem_dtype.type_num, data)
-    else:
-        # dtypes like string & void need a size specified, so can't be used with
-        # SimpleNewFromData. Cython doesn't expose NumPy C-API functions
-        # like NewFromDescr, so we'll construct this with a Python function.
+    elif elem_dtype.kind != b'V':
```

Comment:
I'm a bit confused here - the comment suggests that the `else` branch you're adding is for object arrays, which I can see the sense behind, but object dtypes appear to have `dt.kind == 'O'` rather than 'V'?

Reply:
Yeah, I agree this is confusing. The 'V' data type is actually what numpy uses for structured types (compounds), but if you try to use `np.frombuffer` with 'V' you get the error message "cannot create an OBJECT array from memory buffer".
I've removed the `np.frombuffer` code path since I could not manage to write a test that reached it.

Reply:
I still care about the `np.frombuffer` thing, because the purpose of that is to avoid allocating a new chunk of memory and copying the data across.

> I've removed the np.frombuffer code path since I could not manage to write a test that reached it.

If I haven't missed something in your code, you can hit it by creating a dataset where each entry holds a variable number of void elements:

```
>>> vdt = h5py.vlen_dtype(np.dtype('V4'))
>>> ds = f.create_dataset('v4', dtype=vdt, shape=10)
>>> ds[0] = np.array([b'abcd', b'1234'], dtype='V4')  # Write some data
>>> ds[0]  # Read it back, triggering conversion
```

You can do something similar with numpy string data (which means bytes, in Python 3 terms):

```
>>> ds2 = f.create_dataset('s3', dtype=vdt2, shape=10)
>>> ds2[0] = np.array([b'abc', b'123'], dtype='S3')
>>> ds[0]
```

I think compound dtypes also use a 'V' type, and it looks like there are tests already in the repo that cover that.

> if you try to use np.frombuffer with 'V' you get the error message "cannot create an OBJECT array from memory buffer".

I've just tried, and I get that message with dtype `'O'`, which is what I'd expect, but not with a V dtype (these need a size, like `'V4'`)